semgrep-core: Add metavariable-patttern operator

IagoAbal opened the pull request:
This new (semgrep-core only) operator allows filtering results based on
the program fragment matched by a metavariable, like metavariable-regex
but using an arbitrary pattern formula!

test plan:
make test # tests included

PR checklist:
- [ ] changelog is up to date

IagoAbal commented:
First I did it for just one `pattern-not`. I could add an arbitrary number of `pattern-not`s easily and I think this feature would already be valuable. But I'll try having an arbitrary pattern formula first.

aryx commented on semgrep-core/src/engine/Match_rules.ml:
```diff
@@ -370,6 +434,24 @@ let filter_ranges xs cond =
+         | R.CondPattern (mvar, xpat) -> (
+             match List.assoc_opt mvar bindings with
+             | None ->
+                 Printf.eprintf "filter_ranges: CondPattern: %s not found" mvar;
```
You can use the logger

aryx commented on semgrep-core/src/engine/Match_rules.ml:
```diff
@@ -370,6 +434,24 @@ let filter_ranges xs cond =
+                 let ranges =
+                   matches.matches |> List.map match_result_to_range
+                 in
+                 if ranges = [] then true else false )
```
you can put a comment here to say we assume  xpat is a pattern-not

aryx reviewed (approved):
Looks good! Good start.
You can already merge that if you want.

aryx commented on semgrep-core/src/engine/Match_rules.ml:
```diff
@@ -803,7 +755,79 @@ let rec (evaluate_formula : env -> R.formula -> range_with_mvars list) =
+         | R.CondGeneric e ->
+             let env = Eval_generic.bindings_to_env bindings in
+             Eval_generic.eval_bool env e
+         (* todo: would be nice to have CondRegexp also work on
```
wrong place for this comment. Should be above CondRegexp.

aryx commented on semgrep-core/src/engine/Match_rules.ml:
```diff
@@ -803,7 +755,79 @@ let rec (evaluate_formula : env -> R.formula -> range_with_mvars list) =
+             | Some mval -> (
+                 let mast = program_of_mvalue mval in
+                 let lazy_ast_and_errors = lazy (mast, []) in
+                 let xpatterns = xpatterns_in_formula formula in
```
maybe a comment saying "the code below is quite similar to Match_rules.check below; we could maybe factorize things".

IagoAbal commented:
Thanks! I think I managed to make work the recursive case too:

e.g.
```
foo(bar(a + b), 2, 3)
```
and
```
rules:
  - id: test-mvp
    languages:
      - python
    message: Working!
    patterns:
      - pattern: foo($P, ...)
      - metavariable-pattern:
          metavariable: $P
          patterns:
              - pattern: |
                  bar($X)
              - metavariable-pattern:
                  metavariable: $X
                  patterns:
                    - pattern: ... + ...
    severity: WARNING
```

I will go clean up the code (including using the logger and adding some comments as you pointed out) and some tests!

aryx commented on semgrep-core/src/parsing/Parse_rule.ml:
```diff
@@ -320,11 +268,70 @@ let rec parse_formula_new env (x : J.t) : R.formula =
       error "parse_formula_new"
 
-let parse_formula env (x : string * J.t) : R.pformula =
+and parse_formula env (x : string * J.t) : R.pformula =
```
maybe a comment saying "this is now mutually recursive because of metavariable-pattern: which can
contain itself a formula!"

aryx reviewed (approved):
:exploding_head:

aryx commented:
This might help https://github.com/returntocorp/semgrep/issues/3257

aryx commented:
brendon did a few changes in Match_rules.ml to pass around equivalences, so you'll have to merge develop first.

aryx commented on semgrep-core/src/core/Metavariable.ml:
```diff
@@ -88,6 +88,19 @@ let mvalue_to_any = function
   | T x -> G.T x
   | P x -> G.P x
 
+let program_of_mvalue : mvalue -> G.program option = function
```
I would add a comment here "This is used for metavariable-pattern: where we need to transform the content of a metavariable in a program so we can use evaluate_formula on it".

aryx commented on semgrep-core/src/core/Metavariable.ml:
```diff
@@ -88,6 +88,19 @@ let mvalue_to_any = function
+  | Id (id, None) -> Some [ G.exprstmt (G.N (G.Id (id, G.empty_id_info ()))) ]
+  | N x -> Some [ G.exprstmt (G.N x) ]
+  | Ss stmts -> Some stmts
+  | Args _ | T _ | P _ -> None
```
maybe a logger "Args | T | P not handled yet for metavariable-pattern"

aryx commented on semgrep-core/src/core/Rule.ml:
```diff
@@ -285,4 +288,8 @@ let (convert_formula_old : formula_old -> formula) =
 let formula_of_rule r =
   match r.formula with New f -> f | Old oldf -> convert_formula_old oldf
```
you could call formula_of_pformula then here.

aryx commented on semgrep-core/src/engine/Match_rules.ml:
```diff
@@ -129,6 +129,8 @@ type env = {
   pattern_matches : id_to_match_results;
   (* unused for now, but could be passed down for Range.content_at_range *)
   file : Common.filename;
+  rule : R.rule;
```
maybe a comment explaining why we need the rule and xlang in the enviroment (I guess for metavariable-pattern)?

aryx commented on semgrep-core/src/parsing/Parse_rule.ml:
```diff
@@ -228,7 +188,14 @@ let parse_pattern (id, lang) s =
       R.mk_xpat (Spacegrep ast) s
 
-let rec parse_formula_old env (x : string * J.t) : R.formula_old =
+(* This is now mutually recursive because of metavariable-pattern: which can
```
I would have put this comment next to parse_extra

aryx commented on semgrep-core/tests/OTHER/rules/metavar_pattern_nested.yaml:
```diff
@@ -0,0 +1,17 @@
+          patterns:
+              - pattern: |
+                  bar($X)
+              - metavariable-pattern:
```
so good :)

aryx reviewed (approved):
Great stuff.

aryx commented:
Let's merge this!
After that,  I wonder if we could port the examples in https://r2c.quip.com/4vr9AvBjLete/WIP-Sgrep-New-Syntax-Proposals
I'm especially curious whether we can do the  ... when != foo() of coccinelle via   $...STMTS and a metavariable-pattern: pattern-not: ... foo(); ...

IagoAbal committed:
semgrep-core: Add metavariable-patttern operator

This new (semgrep-core only) operator allows filtering results based on
the program fragment matched by a metavariable, like metavariable-regex
but using an arbitrary pattern formula!

test plan:
make test # tests included

aryx commented on semgrep-core/tests/OTHER/rules/metavar_pattern_old_tls_versions.yaml:
```diff
@@ -0,0 +1,45 @@
+        metavariable: $OPTS
+        patterns:
+            - pattern-not: |
+                {secureOptions: $CONST.SSL_OP_NO_TLSv1 | $CONST.SSL_OP_NO_SSLv3 | $CONST.SSL_OP_NO_SSLv2}
```
hmm another use-case for builtin equivalences around commutative operators.
